File: demo/src/main/java/com/EMPhysics/Vector2D.java
package com.EMPhysics;

public class Vector2D {
    private double x;
    private double y;

    public Vector2D(double x, double y) {
        this.x =x;
        this.y=y;
    } 

    public double getX(){
        return x;
    }
    public double getY(){
        return y;
    }
    public void setX(double x){
        this.x = x;
    }
    public void setY(double y){
        this.y = y;
    }
    public Vector2D add(Vector2D other){
        return new Vector2D(this.getX() + other.getX(), this.getY() + other.getY());
    }
    public Vector2D subtract(Vector2D other) {
        return new Vector2D(this.getX() - other.getX(), this.getY() - other.getY());
    }
    public Vector2D scalar_multiply(double scalar){
        return new Vector2D(this.getX() * scalar, this.getY() * scalar);
    }
    public Vector2D normalize() {
        double magnitude = this.magnitude();
        if (magnitude == 0) {
            throw new ArithmeticException("Cannot normalize a zero vector");
        }
        return new Vector2D(this.getX() / magnitude, this.getY() / magnitude);
    }
    public double magnitude (){
        return Math.sqrt(Math.pow(this.getX(), 2) + Math.pow(this.getY(), 2));
    }
    public double dot_product(Vector2D other) {
        return this.getX() * other.getX() + this.getY() * other.getY();
    }
    public double distanceTo(Vector2D other) {
        double dx = this.getX() - other.getX();
        double dy = this.getY() - other.getY();

        return Math.sqrt(dx*dx + dy*dy);
    }
}

File: app/src/simulation/Vector2D.ts
export class Vector2D {
    private x: number;
    private y: number;

    constructor(x: number, y: number) {
        this.x = x;
        this.y = y;
    }

    public getX(): number {
        return this.x;
    }

    public getY(): number {
        return this.y;
    }

    public setX(x: number): void {
        this.x = x;
    }

    public setY(y: number): void {
        this.y = y;
    }

    public add(other: Vector2D): Vector2D {
        return new Vector2D(this.x + other.x, this.y + other.y);
    }

    public subtract(other: Vector2D): Vector2D {
        return new Vector2D(this.x - other.x, this.y - other.y);
    }

    public scalar_multiply(scalar: number): Vector2D {
        return new Vector2D(this.x * scalar, this.y * scalar);
    }

    public magnitude(): number {
        return Math.sqrt(Math.pow(this.x, 2) + Math.pow(this.y, 2));
    }

    public normalize(): Vector2D {
        const magnitude = this.magnitude();
        if (magnitude === 0) {
            throw new Error("Cannot normalize a zero vector");
        }
        return new Vector2D(this.x / magnitude, this.y / magnitude);
    }

    public dot_product(other: Vector2D): number {
        return this.x * other.x + this.y * other.y;
    }

    public distanceTo(other: Vector2D): number {
        const dx = this.x - other.x;
        const dy = this.y - other.y;
        return Math.sqrt(dx * dx + dy * dy);
    }
}